src/Tasks/system.design/stronglytypedresourcebuilder.cs

```diff
@@ -490,41 +481,62 @@ private static string TruncateAndFormatCommentStringForOutput(string commentStri
         [SuppressMessage("Microsoft.Globalization", "CA1303:DoNotPassLiteralsAsLocalizedParameters")]
         private static bool DefineResourceFetchingProperty(String propertyName, String resourceName, ResourceData data, CodeTypeDeclaration srClass, bool internalClass, bool useStatic)
         {
-            var prop = new CodeMemberProperty
-            {
-                Name = propertyName,
-                HasGet = true,
-                HasSet = false
-            };
+            CodeTypeReference valueType;
+            bool isString;
+            bool isStream;
 
-            Type type = data.Type;
-            if (type == null)
+            if (data.Type is Type type)
             {
-                return false;
-            }
+                // We have an actual object on which we can do type comparisons
+
+                if (type == typeof(MemoryStream))
+                {
+                    type = typeof(UnmanagedMemoryStream);
+                }
+
+                // Ensure type is internalally visible.  This is necessary to ensure
+                // users can access classes via a base type.  Imagine a class like
+                // Image or Stream as a internalally available base class, then an 
+                // internal type like MyBitmap or __UnmanagedMemoryStream as an 
+                // internal implementation for that base class.  For internalally 
+                // available strongly typed resource classes, we must return the 
+                // internal type.  For simplicity, we'll do that for internal strongly 
+                // typed resource classes as well.  Ideally we'd also like to check
+                // for interfaces like IList, but I don't know how to do that without
+                // special casing collection interfaces & ignoring serialization 
+                // interfaces or IDisposable.
+                while (!type.IsPublic)
+                {
+                    type = type.BaseType;
+                }
+
+                valueType = new CodeTypeReference(type);
 
-            if (type == typeof(MemoryStream))
+                isString = type == typeof(String);
+                isStream = type == typeof(UnmanagedMemoryStream) || type == typeof(MemoryStream);
+
+            }
+            else
             {
-                type = typeof(UnmanagedMemoryStream);
+                // We don't have access to a live copy of the object, so
+                // we must do what we can with string information.
+
+                // TODO: can we do the type gymnastics here, too?
+
+                valueType = new CodeTypeReference(data.TypeName);
+
+                // TODO: can we do better here than exact comparisons against something that may vary over time?
```

Comment:
@ericstj I'd love your input on this aspect in particular. Right now it results in generated code like

```c#
    /// <summary>
    ///   Looks up a localized resource of type System.String.
    /// </summary>
    internal static string String1 {
        get {
            object obj = ResourceManager.GetObject("String1", resourceCulture);
            return ((string)(obj));
        }
    }
```

Instead of using `GetString`. I think I can sidestep this instance of the problem by making `StringResource` inherit from `LiveObjectResource` so it falls into that codepath, but I don't have a clear picture what to do about streams.

Reply:
Is this different than what it generated before?  Calling `GetObject` instead of `GetString` is a little concerning as `GetString` is a separate codepath.  I haven't dug into the details to understand concretely if there's functional significance, but this difference is concerning enough that I bet its noticeable.  /cc @tarekms

`GetStream` is less concerning as this just builds on top of `GetObject` I don't see this as a big problem.

Reply:
Yes. In the old days and with my latest push 06ce897df, it generates:

```c#
    /// <summary>
    ///   Looks up a localized string similar to Value1.
    /// </summary>
    internal static string String1 {
        get {
            return ResourceManager.GetString("String1", resourceCulture);
        }
    }
```

If `string` is the more worrisome case, I think it's fixed.

Reply:
So I took a closer look at GetStream and found that it does do something special.
https://github.com/dotnet/corefx/blob/46d3f57a9bba6c79ee494996d465e33fd8effa32/src/Common/src/CoreLib/System/Resources/ResourceManager.cs#L753-L760

It does just call GetObject but it doesn't wrap the underlying UnmanagedMemoryStream.

Contrast this to calling GetObject and casting, and it will wrap the UMS and prevent access to the backing pointer:
https://github.com/dotnet/corefx/blob/46d3f57a9bba6c79ee494996d465e33fd8effa32/src/Common/src/CoreLib/System/Resources/ResourceManager.cs#L703

This would break the scenario where someone is down-casting back to UMS to get the pointer to an embedded stream.  So I think we do need to preserve calls to GetStream too if we can.

Reply:
@ericstj Would you be comfortable with plumbing the assembly qualified type name down to here and doing matches on that?

IIUC I'd need to compare to either the full or core identities, right? so:

```
string:

System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
System.String, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e

MemoryStream:

System.IO.MemoryStream, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
System.IO.MemoryStream, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e

UMS:

System.IO.UnmanagedMemoryStream, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
System.IO.UnmanagedMemoryStream, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
```

And there won't be a way to do the chain-up-to-a-public-type dance at all.